File: ostium-python-sdk/ostium_python_sdk/sdk.py
from dotenv import load_dotenv
import os
from decimal import Decimal, ROUND_DOWN

from ostium_python_sdk.formulae import GetFundingRate
from ostium_python_sdk.scscript.funding import getTargetFundingRate
from ostium_python_sdk.utils import calculate_fee_per_hours, format_with_precision

from .formulae_wrapper import get_trade_metrics
from .constants import CHAIN_ID_ARBITRUM_MAINNET, CHAIN_ID_ARBITRUM_TESTNET, PRECISION_2, PRECISION_6, PRECISION_12, PRECISION_18, PRECISION_9

from ostium_python_sdk.faucet import Faucet
from .balance import Balance
from .price import Price
from web3 import Web3
from .ostium import Ostium
from .config import NetworkConfig
from typing import Union
from .subgraph import SubgraphClient


class OstiumSDK:
    def __init__(self, network: Union[str, NetworkConfig], private_key: str = None, rpc_url: str = None, verbose=False, use_delegation=False):
        self.verbose = verbose
        load_dotenv()
        self.private_key = private_key or os.getenv('PRIVATE_KEY')
        self.use_delegation = use_delegation

        self.rpc_url = rpc_url or os.getenv('RPC_URL')
        if not self.rpc_url:
            network_name = "mainnet" if isinstance(
                network, str) and network == "mainnet" else "testnet"
            raise ValueError(
                f"No RPC_URL provided for {network_name}. Please provide via constructor or RPC_URL environment variable")

        # Initialize Web3
        self.w3 = Web3(Web3.HTTPProvider(self.rpc_url))

        # Get network configuration
        if isinstance(network, NetworkConfig):
            self.network_config = network
        elif isinstance(network, str):
            if network == "mainnet":
                self.network_config = NetworkConfig.mainnet()
            elif network == "testnet":
                self.network_config = NetworkConfig.testnet()
            else:
                raise ValueError(
                    f"Unsupported network: {network}. Use 'mainnet' or 'testnet'")
        else:
            raise ValueError(
                "Network must be either a NetworkConfig instance or a string ('mainnet' or 'testnet')")

        # Validate chain ID
        expected_chain_id = CHAIN_ID_ARBITRUM_MAINNET if not self.network_config.is_testnet else CHAIN_ID_ARBITRUM_TESTNET
        actual_chain_id = self.w3.eth.chain_id
        if actual_chain_id != expected_chain_id:
            raise ValueError(
                f"Chain ID mismatch. Expected {expected_chain_id} for {'testnet' if self.network_config.is_testnet else 'mainnet'}, "
                f"but RPC is connected to chain ID {actual_chain_id}. Please check your RPC_URL."
            )

        if self.verbose:
            print(
                f"network_config: {'TESTNET' if self.network_config.is_testnet else 'MAINNET'}")

        # Initialize Ostium instance
        self.ostium = Ostium(
            self.w3,
            self.network_config.contracts["usdc"],
            self.network_config.contracts["tradingStorage"],
            self.network_config.contracts["trading"],
            private_key=self.private_key,
            verbose=self.verbose,
            use_delegation=self.use_delegation
        )

        # Initialize subgraph client
        self.subgraph = SubgraphClient(
            url=self.network_config.graph_url, verbose=self.verbose)

        self.balance = Balance(
            self.w3, self.network_config.contracts["usdc"], verbose=self.verbose)
        self.price = Price(verbose=self.verbose)

        if self.network_config.is_testnet:
            self.faucet = Faucet(self.w3, self.private_key,
                                 verbose=self.verbose)
        else:
            self.faucet = None

    def log(self, message):
        if self.verbose:
            print(message)

    async def get_open_trades(self, trader_address=None):
        if trader_address is None:
            trader_public_address = self.ostium.get_public_address()
        else:
            trader_public_address = trader_address

        self.log(f"Trader public address: {trader_public_address}")
        open_trades = await self.subgraph.get_open_trades(trader_public_address)
        return open_trades, trader_public_address

    # if SDK instantiated with a private key, this function will return a given open trade metrics,
    # such as: funding fee, roll over fee, Unrealized Pnl, Profit Percent, etc.
    #
    # Will thorw in case SDK instantiated with no private key
    async def get_open_trade_metrics(self, pair_id, trade_index, trader_address=None):
        open_trades, trader_public_address = await self.get_open_trades(trader_address)

        liq_margin_threshold_p = await self.subgraph.get_liq_margin_threshold_p()
        self.log(
            f"SDK: get_open_trade_metrics: {liq_margin_threshold_p}, will use it for liquidation price calculation - call to get_trade_metrics()")

        trade_details = None

        if len(open_trades) == 0:
            raise ValueError(f"No Open Trades for {trader_public_address}")

        for t in open_trades:
            if int(t['pair']['id']) == int(pair_id) and int(t['index']) == int(trade_index):
                trade_details = t
                break

        if trade_details is None:
            raise ValueError(
                f"Trade not found for {trader_public_address} pair {pair_id} and index {trade_index}")

        self.log(f"\nTrade details: {trade_details}")
        # get the price for this trade's asset/feed
        price_data = await self.price.get_latest_price_json(trade_details['pair']['from'], trade_details['pair']['to'])
        self.log(
            f"\nPrice data: {price_data} (contains bid, mid, ask prices among other things)")
        # get the block number
        block_number = self.ostium.get_block_number()
        self.log(f"\nBlock number: {block_number}")

        pair_max_leverage = await self.get_pair_max_leverage(trade_details['pair']['id'])

        return get_trade_metrics(trade_details, price_data, block_number, pair_max_leverage, liq_margin_threshold_p, verbose=self.verbose)

    async def get_target_funding_rate(self, pair_id):
        pair_details = await self.subgraph.get_pair_details(pair_id)

        hillInflectionPoint = Decimal(
            pair_details['hillInflectionPoint']) / PRECISION_18

        maxFundingFeePerBlock = Decimal(
            pair_details['maxFundingFeePerBlock']) / PRECISION_18

        hillPosScale = Decimal(pair_details['hillPosScale']) / PRECISION_2
        hillNegScale = Decimal(pair_details['hillNegScale']) / PRECISION_2

        currLongOI = Decimal(pair_details['longOI']) / PRECISION_6
        currShortOI = Decimal(pair_details['shortOI']) / PRECISION_6
        maxOI = Decimal(pair_details['maxOI']) / PRECISION_6  # maxOI is in USD

        openInterestMax = max(currLongOI, currShortOI)
        normalizedOiDelta = ((currLongOI - currShortOI).quantize(PRECISION_6, rounding=ROUND_DOWN) / max(
            maxOI, openInterestMax).quantize(PRECISION_6, rounding=ROUND_DOWN)).quantize(PRECISION_6, rounding=ROUND_DOWN)

        targetFundingRate = getTargetFundingRate(
            normalizedOiDelta,
            hillInflectionPoint,
            maxFundingFeePerBlock,
            hillPosScale,
            hillNegScale
        )

        self.log(
            f"{pair_details['from']}{pair_details['to']} Traget Funding rate: {targetFundingRate}\n\nPair details: {pair_details}")

        return targetFundingRate

    # max leverage for overnight trades (Stocks) - 100 means 100x, None if not set
    async def get_pair_overnight_max_leverage(self, pair_id):
        obj = await self.subgraph.get_pair_details(pair_id)

        maxLeverage = int(obj['overnightMaxLeverage'])/PRECISION_2 if int(
            obj['overnightMaxLeverage']) != 0 else None
        return maxLeverage

    # either by group of pair or by pair id (e.g: maxLeverage 100 means 100x)
    async def get_pair_max_leverage(self, pair_id):
        obj = await self.subgraph.get_pair_details(pair_id)

        maxLeverage = int(obj['maxLeverage']) / PRECISION_2 if int(
            obj['group']['maxLeverage']) == 0 else int(obj['group']['maxLeverage']) / PRECISION_2
        return maxLeverage

    async def get_pair_net_rate_percent_per_hours(self, pair_id, period_hours=24):
        raise RuntimeError(
            f"Old version of function. Use get_funding_rate_for_pair_id(pair_id, period_hours=24).")

    async def get_rollover_rate_for_pair_id(self, pair_id, period_hours=24):
        pair_details = await self.subgraph.get_pair_details(pair_id)
        rollover_fee_per_block = Decimal(
            pair_details['rolloverFeePerBlock']) / Decimal('1e18')
        rollover = calculate_fee_per_hours(
            rollover_fee_per_block, hours=period_hours)
        return rollover

    async def get_funding_rate_for_pair_id(self, pair_id, period_hours=24):
        pair_details = await self.subgraph.get_pair_details(pair_id)
        # get the block number
        block_number = self.ostium.get_block_number()

        # Get current price
        last_trade_price = pair_details['lastTradePrice']

        long_oi = int(
            (Decimal(pair_details['longOI']) *
             Decimal(last_trade_price) / PRECISION_18 / PRECISION_12)
        )
        short_oi = int(
            (Decimal(pair_details['shortOI']) *
             Decimal(last_trade_price) / PRECISION_18 / PRECISION_12)
        )

        ret = GetFundingRate(
            pair_details['accFundingLong'],
            pair_details['accFundingShort'],
            pair_details['lastFundingRate'],
            pair_details['maxFundingFeePerBlock'],
            pair_details['lastFundingBlock'],
            block_number,
            str(long_oi),  # Needs to be in USD
            str(short_oi),  # Needs to be in USD
            str(pair_details['maxOI']),
            pair_details['hillInflectionPoint'],
            pair_details['hillPosScale'],
            pair_details['hillNegScale'],
            pair_details['springFactor'],
            pair_details['sFactorUpScaleP'],
            pair_details['sFactorDownScaleP'],
            self.verbose
        )

        accFundingLong = ret['accFundingLong']
        accFundingShort = ret['accFundingShort']
        fundingRate = float(ret['latestFundingRate']) * \
            (10 / 3) * 60 * 60 * 100 * period_hours
        targetFundingRate = float(ret['targetFundingRate']) * \
            (10 / 3) * 60 * 60 * 100 * period_hours

        self.log(
            f"{pair_details['from']}{pair_details['to']} Funding rate ({period_hours} hours): {fundingRate}%")
        self.log(
            f"{pair_details['from']}{pair_details['to']} Traget Funding rate ({period_hours} hours): {targetFundingRate}%")

        return accFundingLong, accFundingShort, fundingRate, targetFundingRate

    async def get_formatted_pairs_details(self, including_current_price_and_market_status=True) -> list:
        pairs = await self.subgraph.get_pairs()
        formatted_pairs = []

        for pair in pairs:
            formatted_pair = {
                'id': int(pair['id']),
                'from': pair['from'],
                'to': pair['to'],
                'group': pair['group']['name'],
                'longOI': Decimal(pair['longOI']) / PRECISION_18,
                'shortOI': Decimal(pair['shortOI']) / PRECISION_18,
                'maxOI': Decimal(pair['maxOI']) / PRECISION_6,
                'makerFeeP': Decimal(pair['makerFeeP']) / PRECISION_6,
                'takerFeeP': Decimal(pair['takerFeeP']) / PRECISION_6,
                'minLeverage': int(pair['minLeverage']) / PRECISION_2 if int(
                    pair['group']['minLeverage']) == 0 else Decimal(pair['group']['minLeverage']) / PRECISION_2,
                'maxLeverage': int(pair['maxLeverage']) / PRECISION_2 if int(
                    pair['group']['maxLeverage']) == 0 else Decimal(pair['group']['maxLeverage']) / PRECISION_2,
                'makerMaxLeverage': Decimal(pair['makerMaxLeverage']) / PRECISION_2,
                'groupMaxCollateralP': Decimal(pair['group']['maxCollateralP']) / PRECISION_2,
                'minLevPos': Decimal(pair['fee']['minLevPos']) / PRECISION_6,
                'lastFundingRate': Decimal(pair['lastFundingRate']) / PRECISION_9,
                'curFundingLong': Decimal(pair['curFundingLong']) / PRECISION_9,
                'curFundingShort': Decimal(pair['curFundingShort']) / PRECISION_9,
                'lastFundingBlock': int(pair['lastFundingBlock'])
            }

            if int(pair['overnightMaxLeverage']) != 0:
                formatted_pair['overnightMaxLeverage'] = Decimal(
                    pair['overnightMaxLeverage']) / PRECISION_2

            if including_current_price_and_market_status:
                # Get current price and market status
                try:
                    price, is_market_open, is_day_trading_closed = await self.price.get_price(
                        pair['from'],
                        pair['to']
                    )
                    if price is not None:
                        formatted_pair['price'] = price
                    if is_market_open is not None:
                        formatted_pair['isMarketOpen'] = is_market_open
                    if is_day_trading_closed is not None:
                        formatted_pair['isDayTradingClosed'] = is_day_trading_closed
                except ValueError:
                    pass

            formatted_pairs.append(formatted_pair)

        formatted_pairs.sort(key=lambda x: x['id'])
        return formatted_pairs

File: src/sdk.rs
//! Top-level [`OstiumSDK`] and its [`OstiumSDKBuilder`].

use std::str::FromStr;

use alloy::primitives::Address;
use alloy::providers::{DynProvider, Provider, ProviderBuilder};
use alloy::signers::local::PrivateKeySigner;
use rust_decimal::prelude::ToPrimitive;
use rust_decimal::{Decimal, RoundingStrategy};
use url::Url;

use crate::config::NetworkConfig;
use crate::consts::{
    CHAIN_ID_ARB_MAINNET, CHAIN_ID_ARB_TESTNET, PRECISION_2, PRECISION_6, PRECISION_9,
    PRECISION_12, PRECISION_18,
};
use crate::error::{OstiumError, Result};
use crate::faucet::Faucet;
use crate::formulae::funding::get_target_funding_rate;
use crate::formulae::get_funding_rate;
use crate::formulae::{TradeMetrics, get_trade_metrics};
use crate::ostium::Ostium;
use crate::price::Price;
use crate::subgraph::{DecExt, DecOptExt, OpenTrade, SubgraphClient};
use crate::utils::calculate_fee_per_hours;

/// Network selector — pick mainnet/testnet by enum or pass a custom config.
#[derive(Debug, Clone)]
pub enum Network {
    Mainnet,
    Testnet,
    Custom(NetworkConfig),
}

/// Backwards-compat type alias matching pre-refactor naming.
pub type NetworkSpec = Network;

impl<'a> std::convert::TryFrom<&'a str> for Network {
    type Error = OstiumError;
    fn try_from(s: &'a str) -> Result<Self> {
        match s {
            "mainnet" => Ok(Network::Mainnet),
            "testnet" => Ok(Network::Testnet),
            other => Err(OstiumError::UnsupportedNetwork(other.to_string())),
        }
    }
}

impl From<NetworkConfig> for Network {
    fn from(cfg: NetworkConfig) -> Self {
        Network::Custom(cfg)
    }
}

impl Network {
    fn name(&self) -> &'static str {
        match self {
            Network::Mainnet => "mainnet",
            Network::Testnet => "testnet",
            Network::Custom(c) => {
                if c.is_testnet {
                    "testnet"
                } else {
                    "mainnet"
                }
            }
        }
    }

    fn into_config(self) -> NetworkConfig {
        match self {
            Network::Mainnet => NetworkConfig::mainnet(),
            Network::Testnet => NetworkConfig::testnet(),
            Network::Custom(c) => c,
        }
    }
}

/// Top-level SDK aggregating the on-chain client, subgraph, price feed, and
/// optional faucet.
pub struct OstiumSDK {
    pub use_delegation: bool,
    pub rpc_url: String,
    pub network_config: NetworkConfig,
    pub provider: DynProvider,
    pub ostium: Ostium,
    pub subgraph: SubgraphClient,
    pub price: Price,
    /// Only set when running on testnet.
    pub faucet: Option<Faucet>,
}

impl OstiumSDK {
    /// Begin constructing an [`OstiumSDK`].
    pub fn builder() -> OstiumSDKBuilder {
        OstiumSDKBuilder::default()
    }

    /// Get open trades for `trader`, defaulting to the SDK's wallet address.
    pub async fn get_open_trades(
        &self,
        trader_address: Option<Address>,
    ) -> Result<(Vec<OpenTrade>, Address)> {
        let trader = match trader_address {
            Some(a) => a,
            None => self.ostium.public_address()?,
        };
        tracing::debug!("trader public address: {:?}", trader);
        let trades = self.subgraph.get_open_trades(&format!("{:?}", trader)).await?;
        Ok((trades, trader))
    }

    /// Compute live metrics for a single open trade (PnL, funding, rollover,
    /// liquidation price).
    pub async fn get_open_trade_metrics(
        &self,
        pair_id: u64,
        trade_index: u64,
        trader_address: Option<Address>,
    ) -> Result<TradeMetrics> {
        let (open_trades, trader) = self.get_open_trades(trader_address).await?;

        let liq_margin_threshold_p = self.subgraph.get_liq_margin_threshold_p().await?;
        tracing::debug!(
            "liq_margin_threshold_p: {} (used for liquidation price)",
            liq_margin_threshold_p
        );

        if open_trades.is_empty() {
            return Err(OstiumError::NoOpenTrades(trader));
        }

        let trade = open_trades
            .into_iter()
            .find(|t| {
                t.pair.id.parse::<u64>().ok() == Some(pair_id)
                    && t.index.parse::<u64>().ok() == Some(trade_index)
            })
            .ok_or(OstiumError::TradeNotFound {
                trader,
                pair_id,
                index: trade_index,
            })?;

        let price_data = self.price.get_latest_price(&trade.pair.from, &trade.pair.to).await?;
        let block_number = self.ostium.block_number().await?;
        let pair_max_leverage = self.get_pair_max_leverage(&trade.pair.id).await?;

        Ok(get_trade_metrics(
            &trade,
            &price_data,
            block_number,
            pair_max_leverage,
            liq_margin_threshold_p,
        ))
    }

    /// Compute the target funding rate for a pair.
    pub async fn get_target_funding_rate(&self, pair_id: &str) -> Result<Decimal> {
        let pair = self.subgraph.get_pair_details(pair_id).await?;

        let hill_inflection_point = pair.hill_inflection_point.dec() / PRECISION_18;
        let max_funding_fee_per_block = pair.max_funding_fee_per_block.dec() / PRECISION_18;
        let hill_pos_scale = pair.hill_pos_scale.dec() / PRECISION_2;
        let hill_neg_scale = pair.hill_neg_scale.dec() / PRECISION_2;

        let curr_long_oi = pair.long_oi.dec() / PRECISION_6;
        let curr_short_oi = pair.short_oi.dec() / PRECISION_6;
        let max_oi = pair.max_oi.dec() / PRECISION_6;

        let open_interest_max = curr_long_oi.max(curr_short_oi);
        let denom = max_oi
            .max(open_interest_max)
            .round_dp_with_strategy(6, RoundingStrategy::ToZero);
        let normalized_oi_delta = ((curr_long_oi - curr_short_oi)
            .round_dp_with_strategy(6, RoundingStrategy::ToZero)
            / denom)
            .round_dp_with_strategy(6, RoundingStrategy::ToZero);

        let target = get_target_funding_rate(
            normalized_oi_delta,
            hill_inflection_point,
            max_funding_fee_per_block,
            hill_pos_scale,
            hill_neg_scale,
        );

        tracing::debug!(
            "{}{} target funding rate: {}",
            pair.from,
            pair.to,
            target
        );

        Ok(target)
    }

    /// Maximum leverage allowed for overnight (stock-style) trades, if set.
    pub async fn get_pair_overnight_max_leverage(
        &self,
        pair_id: &str,
    ) -> Result<Option<Decimal>> {
        let pair = self.subgraph.get_pair_details(pair_id).await?;
        Ok(pair
            .overnight_max_leverage
            .dec_opt()
            .filter(|v| !v.is_zero())
            .map(|v| v / PRECISION_2))
    }

    /// Maximum leverage for the pair, falling back to the group's max.
    pub async fn get_pair_max_leverage(&self, pair_id: &str) -> Result<Decimal> {
        let pair = self.subgraph.get_pair_details(pair_id).await?;
        let group_max = pair.group.max_leverage.dec();
        Ok(if group_max.is_zero() {
            pair.max_leverage.dec() / PRECISION_2
        } else {
            group_max / PRECISION_2
        })
    }

    /// Rollover fee accrual rate over `period_hours`.
    pub async fn get_rollover_rate_for_pair_id(
        &self,
        pair_id: &str,
        period_hours: u64,
    ) -> Result<Decimal> {
        let pair = self.subgraph.get_pair_details(pair_id).await?;
        let rollover_fee_per_block = pair.rollover_fee_per_block.dec() / PRECISION_18;
        Ok(calculate_fee_per_hours(rollover_fee_per_block, period_hours, 5))
    }

    /// `(acc_funding_long, acc_funding_short, funding_rate, target_funding_rate)`
    /// scaled to `period_hours`.
    pub async fn get_funding_rate_for_pair_id(
        &self,
        pair_id: &str,
        period_hours: u64,
    ) -> Result<(Decimal, Decimal, f64, f64)> {
        let pair = self.subgraph.get_pair_details(pair_id).await?;
        let block_number = self.ostium.block_number().await?;

        let last_trade_price = pair.last_trade_price.dec();
        let long_oi = pair.long_oi.dec();
        let short_oi = pair.short_oi.dec();

        let long_oi_calc = (long_oi * last_trade_price) / PRECISION_18 / PRECISION_12;
        let short_oi_calc = (short_oi * last_trade_price) / PRECISION_18 / PRECISION_12;
        let long_oi_int = long_oi_calc
            .round_dp_with_strategy(0, RoundingStrategy::ToZero)
            .to_string();
        let short_oi_int = short_oi_calc
            .round_dp_with_strategy(0, RoundingStrategy::ToZero)
            .to_string();

        let ret = get_funding_rate(
            pair.acc_funding_long.dec(),
            pair.acc_funding_short.dec(),
            pair.last_funding_rate.dec(),
            pair.max_funding_fee_per_block.dec(),
            pair.last_funding_block.dec(),
            Decimal::from(block_number),
            Decimal::from_str(&long_oi_int).map_err(|e| OstiumError::Conversion(e.to_string()))?,
            Decimal::from_str(&short_oi_int).map_err(|e| OstiumError::Conversion(e.to_string()))?,
            pair.max_oi.dec(),
            pair.hill_inflection_point.dec(),
            pair.hill_pos_scale.dec(),
            pair.hill_neg_scale.dec(),
            pair.spring_factor.dec(),
            pair.s_factor_up_scale_p.dec(),
            pair.s_factor_down_scale_p.dec(),
        );

        let multiplier = (10.0_f64 / 3.0) * 60.0 * 60.0 * 100.0 * period_hours as f64;
        let funding_rate = ret.latest_funding_rate.to_f64().unwrap_or(0.0) * multiplier;
        let target_funding_rate = ret.target_funding_rate.to_f64().unwrap_or(0.0) * multiplier;

        tracing::debug!(
            "{}{} funding rate ({}h): {}% target: {}%",
            pair.from,
            pair.to,
            period_hours,
            funding_rate,
            target_funding_rate
        );

        Ok((
            ret.acc_funding_long,
            ret.acc_funding_short,
            funding_rate,
            target_funding_rate,
        ))
    }

    /// Scaled, sorted snapshot of every pair, optionally augmented with the
    /// current price feed.
    pub async fn get_formatted_pairs_details(
        &self,
        including_current_price_and_market_status: bool,
    ) -> Result<Vec<FormattedPair>> {
        let pairs = self.subgraph.get_pairs().await?;
        let mut formatted: Vec<FormattedPair> = Vec::with_capacity(pairs.len());

        for pair in pairs.into_iter() {
            let group_max = pair.group.max_leverage.dec();
            let max_leverage = if group_max.is_zero() {
                pair.max_leverage.dec() / PRECISION_2
            } else {
                group_max / PRECISION_2
            };
            let overnight = pair
                .overnight_max_leverage
                .dec_opt()
                .filter(|v| !v.is_zero())
                .map(|v| v / PRECISION_2);

            let mut entry = FormattedPair {
                id: pair.id.parse().unwrap_or(0),
                from: pair.from.clone(),
                to: pair.to.clone(),
                group: pair.group.name,
                long_oi: pair.long_oi.dec() / PRECISION_18,
                short_oi: pair.short_oi.dec() / PRECISION_18,
                max_oi: pair.max_oi.dec() / PRECISION_6,
                maker_fee_p: pair.maker_fee_p.dec() / PRECISION_6,
                taker_fee_p: pair.taker_fee_p.dec() / PRECISION_6,
                max_leverage,
                maker_max_leverage: pair.maker_max_leverage.dec() / PRECISION_2,
                group_max_collateral_p: pair.group.max_collateral_p.dec() / PRECISION_2,
                min_lev_pos: pair.fee.min_lev_pos.dec() / PRECISION_6,
                last_funding_rate: pair.last_funding_rate.dec() / PRECISION_9,
                cur_funding_long: pair.cur_funding_long.dec() / PRECISION_9,
                cur_funding_short: pair.cur_funding_short.dec() / PRECISION_9,
                last_funding_block: pair.last_funding_block.parse().unwrap_or(0),
                overnight_max_leverage: overnight,
                price: None,
                is_market_open: None,
                is_day_trading_closed: None,
            };

            if including_current_price_and_market_status {
                if let Ok((price, is_market_open, is_day_trading_closed)) =
                    self.price.get_price(&pair.from, &pair.to).await
                {
                    entry.price = Some(price);
                    entry.is_market_open = Some(is_market_open);
                    entry.is_day_trading_closed = Some(is_day_trading_closed);
                }
            }

            formatted.push(entry);
        }

        formatted.sort_by_key(|p| p.id);
        Ok(formatted)
    }
}

/// Per-pair snapshot returned by [`OstiumSDK::get_formatted_pairs_details`].
#[derive(Debug, Clone)]
pub struct FormattedPair {
    pub id: u64,
    pub from: String,
    pub to: String,
    pub group: String,
    pub long_oi: Decimal,
    pub short_oi: Decimal,
    pub max_oi: Decimal,
    pub maker_fee_p: Decimal,
    pub taker_fee_p: Decimal,
    pub max_leverage: Decimal,
    pub maker_max_leverage: Decimal,
    pub group_max_collateral_p: Decimal,
    pub min_lev_pos: Decimal,
    pub last_funding_rate: Decimal,
    pub cur_funding_long: Decimal,
    pub cur_funding_short: Decimal,
    pub last_funding_block: u64,
    pub overnight_max_leverage: Option<Decimal>,
    pub price: Option<f64>,
    pub is_market_open: Option<bool>,
    pub is_day_trading_closed: Option<bool>,
}

/// Builder for [`OstiumSDK`]. Required: [`network`](Self::network) and
/// [`rpc_url`](Self::rpc_url) (the latter falls back to the `RPC_URL`
/// environment variable).
#[derive(Default)]
pub struct OstiumSDKBuilder {
    network: Option<Network>,
    rpc_url: Option<String>,
    private_key: Option<String>,
    signer: Option<PrivateKeySigner>,
    use_delegation: bool,
}

impl OstiumSDKBuilder {
    pub fn network(mut self, network: impl Into<Network>) -> Self {
        self.network = Some(network.into());
        self
    }

    pub fn rpc_url(mut self, rpc_url: String) -> Self {
        self.rpc_url = Some(rpc_url);
        self
    }

    /// Set a private key as a hex string. Convenience for env-var driven setups.
    pub fn private_key_hex(mut self, pk: impl Into<String>) -> Self {
        self.private_key = Some(pk.into());
        self
    }

    /// Set the signer directly.
    pub fn signer(mut self, signer: PrivateKeySigner) -> Self {
        self.signer = Some(signer);
        self
    }

    pub fn use_delegation(mut self, v: bool) -> Self {
        self.use_delegation = v;
        self
    }

    pub async fn build(self) -> Result<OstiumSDK> {
        let network = self
            .network
            .ok_or(OstiumError::Builder("missing network"))?;
        let rpc_url = self
            .rpc_url
            .or_else(|| std::env::var("RPC_URL").ok())
            .ok_or_else(|| OstiumError::MissingRpcUrl {
                network: match &network {
                    Network::Mainnet => "mainnet",
                    Network::Testnet => "testnet",
                    Network::Custom(c) => {
                        if c.is_testnet {
                            "testnet"
                        } else {
                            "mainnet"
                        }
                    }
                },
            })?;

        let signer = match (self.signer, self.private_key) {
            (Some(s), _) => Some(s),
            (None, Some(hex)) => Some(
                PrivateKeySigner::from_str(&hex)
                    .map_err(|e| OstiumError::InvalidPrivateKey(e.to_string()))?,
            ),
            (None, None) => match std::env::var("PRIVATE_KEY") {
                Ok(hex) => Some(
                    PrivateKeySigner::from_str(&hex)
                        .map_err(|e| OstiumError::InvalidPrivateKey(e.to_string()))?,
                ),
                Err(_) => None,
            },
        };

        let net_name = network.name();
        let network_config = network.into_config();

        let url = Url::parse(&rpc_url)?;
        let provider = ProviderBuilder::new().connect_http(url).erased();

        let actual = provider
            .get_chain_id()
            .await
            .map_err(|e| OstiumError::Contract(e.to_string()))?;
        let expected = if network_config.is_testnet {
            CHAIN_ID_ARB_TESTNET
        } else {
            CHAIN_ID_ARB_MAINNET
        };
        if actual != expected {
            return Err(OstiumError::ChainIdMismatch {
                expected,
                actual,
                network: if network_config.is_testnet {
                    "testnet"
                } else {
                    "mainnet"
                },
            });
        }

        tracing::debug!(
            "network: {} (chain id {}, RPC {}, name {})",
            if network_config.is_testnet { "TESTNET" } else { "MAINNET" },
            actual,
            rpc_url,
            net_name
        );

        let ostium = Ostium::new(
            provider.clone(),
            network_config.contracts.usdc,
            network_config.contracts.trading_storage,
            network_config.contracts.trading,
            signer.clone(),
            self.use_delegation,
        );

        let subgraph = SubgraphClient::new(network_config.graph_url.to_string());
        let price = Price::new();
        let faucet = if network_config.is_testnet {
            Some(Faucet::new(provider.clone(), signer))
        } else {
            None
        };

        Ok(OstiumSDK {
            use_delegation: self.use_delegation,
            rpc_url,
            network_config,
            provider,
            ostium,
            subgraph,
            price,
            faucet,
        })
    }
}
